lib/logstash/inputs/s3.rb

```diff
@@ -437,6 +437,23 @@ def get_s3object
     s3 = Aws::S3::Resource.new(options)
   end
 
+  private
+  def file_restored?(object)
```

Comment:
unfortunately we're still assuming header ordering as was with the string previously, had smt like this in mind:
```ruby
def file_restored?(object)
  begin
    restore = object.data.restore
    if restore && restore.match(/ongoing-request\s?=\s?["']false["']/)
      if restore = restore.match(/expiry-date\s?=\s?["'](.*?)["']/)
        expiry_date = DateTime.parse(restore[1])
        return true if expiry_date < Time.now # restored
      else
        return nil # no expiry-date found for ongoing request
      end
    end
  rescue => e
    @logger.debug("Could not determine Glacier restore status.", :exception => e.message)
  end
  return false
end
... also I changed your `expiry_date > Time.now` condition to `>`
I did not test this but as far as I understand if the expiration date is reached than the object is no longer available.
could you please test this with s3?
```

Reply:
I've updated my PR with your suggestions, with a couple of minor corrections.
As I understand it, the file will no longer be available once the expiry time is reached.  I have a test restored file which should be expiring tonight, so I can report back tomorrow once I know whether the file has become unavailable or not.  I assume you wouldn't want Logstash to attempt to download something which is advertised as no longer being available anyway.

I also have a file with a storage class of deep archive currently restoring, so I will soon be able to test this feature against a file with that storage class.  This restore, as you can imagine, takes a bit longer than a standard Glacier restore, but I'll report back if this behaves in the same way once it has completed.

Reply:
cool, please report back. also we might have forgotten to `require 'date'`

Reply:
Will do.  I did wonder about `date`, but from my local testing, it seems to work without being explicitly required (does `time` provide those classes?)

Reply:
The deep glacier handling works as expected:
```
[2020-03-11T09:44:15,083][DEBUG][logstash.inputs.s3       ][main] S3 input: Found key {:key=>"test-deeparchive-restored"}
[2020-03-11T09:44:15,112][INFO ][logstash.inputs.s3       ][main] Using default generated file for the sincedb {:filename=>"/usr/share/logstash/data/plugins/inputs/s3/sincedb_62a0d356ed2b4a5540fe614a028fabf9"}
[2020-03-11T09:44:15,269][DEBUG][logstash.inputs.s3       ][main] S3 input: Adding to objects[] {:key=>"test-deeparchive-restored"}
[2020-03-11T09:44:15,271][DEBUG][logstash.inputs.s3       ][main] objects[] length is:  {:length=>1}
[2020-03-11T09:44:15,359][DEBUG][logstash.inputs.s3       ][main] S3 input: Found key {:key=>"testfile-deeparchive"}
[2020-03-11T09:44:15,441][DEBUG][logstash.inputs.s3       ][main] S3 Input: Object Archived to Glacier {:key=>"testfile-deeparchive"}
```
We can see that the file that has been restored is added to the list, but the one that is still archived is not.  The restored file is then successfully pulled from S3:
```
[2020-03-11T09:44:16,212][DEBUG][logstash.inputs.s3       ][main] Processing file {:filename=>"/tmp/logstash/test-deeparchive-restored"}
{
      "@version" => "1",
    "@timestamp" => 2020-03-11T09:44:16.216Z,
       "message" => "deep\n"
}
```
The file that I mentioned was due to expire is, interestingly, is still downloadable, although officially it has expired.  I expect it will become unavailable in the coming hours.

Reply:
Update: The expired file is no longer downloadable.  I assume there's some sort of deletion job that runs on a schedule and picks up all restored files that are past their expiry date/time.

Reply:
thanks - so all as we expected. 
do you want to bump the version and add a changelog entry ? 
(can do that later if you're unsure about it)

> Will do. I did wonder about date, but from my local testing, it seems to work without being explicitly required (does time provide those classes?)

no `require 'time'` does not load `Date/DateTime`, 
`require 'date'` is loaded by logstash-core, would be nice to have it explicitly here as well.

Reply:
No problem, I've added that requirement and updated the version and changelog.  Hope this all looks OK.